### PR TITLE
Add Scripting API

### PR DESCRIPTION
**Pull request by mdcerna**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-routing-and-faulting-custom-device/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

First submission of the Switch Scripting API and the shared library Switch Shared.lvlib.

### Why should this Pull Request be merged?

This scripting library contains what we need to implement the System Explorer pages needed in the Routing and Faulting CD.

### What testing has been done?

There are a few automated VI tests: deployment of a scripted Routing and Faulting CD, creation of a routing channel with 2 states, and basic tests for connector pane and VI settings.  

**Next up**: more tests to cover all functionality (adding/removing routing channels, states, connections).

### CODE REVIEW

**Comment by rtzoeller** [comment]
Please update the pull request title to something more descriptive than the branch name.

**Comment by csjall** [comment]
- [ ] Remove Value Change virtual folder under Alias. Add, Remove, Modify can all reside at same level.
image
- [ ] Show label on boolean constant. It documents what the constant relates to.
image
- [ ] D.R.Y. Call `Get Alias` from within `Get All Aliases`. Aggregate the result of found.
image
- [ ] D.R.Y. Call `Remove Alias` from within `Remove All Aliases`.
image

**Comment by csjall** [comment]
- [ ] I like the Add RAF channel test. I would like to see some more unit tests for the scripting API. Write positive and negative tests. Confirm that invalid inputs to methods returns appropriate errors.

**Comment by mdcerna** [comment]
All feedback has been implemented, including the creation of a build spec for the scripting API llb.

**Comment by mdcerna** [comment]
> * [x]  The scripting API build spec should output to the `Built` directory like the other build specs.

ok, I changed the path to point to the same Built directory.  On my machine, the full build path is:
     "C:\dev\CustomDevice\niveristand-routing-and-faulting-custom-device\Built\Routing and Faulting\Routing and Faulting Scripting API.llb"

**Comment by buckd** [comment]
> ok, I changed the path to point to the same Built directory. On my machine, the full build path is:
> "C:\dev\CustomDevice\niveristand-routing-and-faulting-custom-device\Built\Routing and Faulting\Routing and Faulting Scripting API.llb"

That's where it gets built on my machine too, which is not where it should since I don't clone to `C:\dev`. It looks like some of your build specs somehow got changed to absolute paths instead of relative paths.

All of the RT build specs correctly build a path relative to the project, but everything under **My Computer** is absolute. I'm not sure how that got changed, but in the project XML, the _Bld_localDestDir_ is absolute and there is no tag for _Bld_localDestDirType_ like there is for the other target build specs.

These paths need to be relative or the automated builds won't archive them correctly.

**Comment by csjall** [comment]
- [ ] Change Scripting Library build specification to copy built libraries into Built/Scripting. This matches the source folder layout.
* Source\Routing and Faulting -> Built\Routing and Faulting\Windows
* Source\SLSC Switch -> Built\SLSC Switch\Windows
* Source\Scripting -> Built\Scripting\Windows

- [ ] LabVIEW saved the build specification to use absolute path instead of relative path. Edit the *.lvproj by hand to fix this (since I have no idea how to do it in the IDE).
```xml
<Item Name="Configuration Release" Type="Source Distribution">
	<Property Name="Bld_buildCacheID" Type="Str">{F7E936A1-C50B-4217-AC2B-A4DA3629CD4B}</Property>
	<Property Name="Bld_buildSpecName" Type="Str">Configuration Release</Property>
	<Property Name="Bld_excludeLibraryItems" Type="Bool">true</Property>
	<Property Name="Bld_excludePolymorphicVIs" Type="Bool">true</Property>
	<Property Name="Bld_excludeTypedefs" Type="Bool">true</Property>
	<Property Name="Bld_localDestDir" Type="Path">../../Built/Routing and Faulting</Property>
	<Property Name="Bld_localDestDirType" Type="Str">relativeToCommon</Property>
	<Property Name="Bld_postActionVIID" Type="Ref">/My Computer/Utility/Copy .LLB to NI VeriStand dir.vi</Property>
	<Property Name="Bld_preActionVIID" Type="Ref">/My Computer/Utility/LLB Pre-Build CHM Build.vi</Property>
	<Property Name="Bld_previewCacheID" Type="Str">{E8035317-FDF9-4FC1-9EF0-1450E8CAE472}</Property>
	<Property Name="Bld_version.major" Type="Int">1</Property>
	<Property Name="Destination[0].destName" Type="Str">Destination Directory</Property>
	<Property Name="Destination[0].path" Type="Path">../../Built/Routing and Faulting</Property>
	<Property Name="Destination[1].destName" Type="Str">Support Directory</Property>
	<Property Name="Destination[1].path" Type="Path">../../Built/Routing and Faulting/Windows/Data</Property>
	<Property Name="Destination[2].destName" Type="Str">Routing and Faulting Configuration LLB</Property>
	<Property Name="Destination[2].path" Type="Path">../../Built/Routing and Faulting/Windows/Routing and Faulting Configuration.llb</Property>
	<Property Name="Destination[2].type" Type="Str">LLB</Property>
	<Property Name="DestinationCount" Type="Int">3</Property>
```

- [ ] Move scripting constants into the Scripting source folder (Source\Scripting\Constants.vi). Currently you are still referencing the constants from the Source\Shared folder.
- [ ] Remove `Switch Shared.lvlib` from the project and the folder structure. We can add it in the future when we add shared components for the UI controls.
image

**Comment by csjall** [comment]
> > ok, I changed the path to point to the same Built directory. On my machine, the full build path is:
> > "C:\dev\CustomDevice\niveristand-routing-and-faulting-custom-device\Built\Routing and Faulting\Routing and Faulting Scripting API.llb"
> 
> That's where it gets built on my machine too, which is not where it should since I don't clone to `C:\dev`. It looks like some of your build specs somehow got changed to absolute paths instead of relative paths.
> 
> All of the RT build specs correctly build a path relative to the project, but everything under **My Computer** is absolute. I'm not sure how that got changed, but in the project XML, the _Bld_localDestDir_ is absolute and there is no tag for _Bld_localDestDirType_ like there is for the other target build specs.
> 
> These paths need to be relative or the automated builds won't archive them correctly.

I have found that LabVIEW will convert relative to absolute path when you save the project. The workaround is to edit the project file by hand and then not change the project properties again...

**Comment by buckd** [comment]
> Change Scripting Library build specification to copy built libraries into Built/Scripting. This matches the source folder layout.

This won't work well with building nipkg installers. It will just put a `Scripting` folder at the top-level of the Custom Devices directory. The scripting stuff needs to be with the rest of the RaF custom device.

File: Source/Routing and Faulting/Routing and Faulting Custom Device.lvproj [hunk 1]
```diff
@@ -682,19 +675,19 @@
...
+				<Property Name="Bld_version.build" Type="Int">6</Property>
 				<Property Name="Bld_version.major" Type="Int">1</Property>
 				<Property Name="Destination[0].destName" Type="Str">Destination Directory</Property>
-				<Property Name="Destination[0].path" Type="Path">../Built/Routing and Faulting/Routing and Faulting Scripting API.llb</Property>
+				<Property Name="Destination[0].path" Type="Path">/C/dev/CustomDevice/niveristand-routing-and-faulting-custom-device/Built/Routing and Faulting/Windows/Routing and Faulting Scripting API.llb</Property>
 				<Property Name="Destination[0].path.type" Type="Str">&lt;none&gt;</Property>
 				<Property Name="Destination[0].type" Type="Str">LLB</Property>
 				<Property Name="Destination[1].destName" Type="Str">Support Directory</Property>
-				<Property Name="Destination[1].path" Type="Path">../Built/Routing and Faulting/data</Property>
+				<Property Name="Destination[1].path" Type="Path">/C/dev/CustomDevice/niveristand-routing-and-faulting-custom-device/Built/Routing and Faulting/Windows/data</Property>
```
Comment (csjall) [inline Source/Routing and Faulting/Routing and Faulting Custom Device.lvproj]: The Scripting Build Spec still has absolute path. Update the destination directories to match other build specifications.

**Comment by csjall** [comment]
I found a runtime issue with the Configuration library. We need to check the option to modify the project library after removing unused members.

image

If we don't the Routing and Faulting custom device will not load in VeriStand.

**Comment by buckd** [comment]
We will also need to have the build change the name of one of the `Constants.vi` files once the one in the Engine lvlib is included in the build. Otherwise, we will hit a name collision. It's not urgent at the moment, but it will be at some point.

**Pull request merged**